foxai/explainer/computer_vision/algorithm/gradient_utils.py

```diff
@@ -0,0 +1,174 @@
+    """
+    if len(signature(forward_func).parameters) == 0:
+        output = forward_func()
+    elif additional_forward_args is not None:
+        output = forward_func((input_tensor, *additional_forward_args))
+    else:
+        output = forward_func(input_tensor)
+
+    if target is None:
+        return output
+    elif isinstance(target, int):
+        return output[:, target]
+    elif isinstance(target, torch.Tensor):
+        assert torch.numel(target) == 1 and isinstance(
+            target.item(), int
+        ), f"Tensor {target} is not a valid class"
+        return output[:, target.item()]
+    else:
+        raise AssertionError(f"Target type {type(target)} is not valid.")
+
+
+def compute_gradients(
+    forward_fn: Callable,
+    input_tensor: torch.Tensor,
+    target_ind: Optional[TargetType] = None,
+    additional_forward_args: Optional[Tuple[Any]] = None,
+) -> torch.Tensor:
+    r"""
+    Computes gradients of the output with respect to inputs for an arbitrary forward function.
+
+    Args:
+        forward_fn: forward function. This can be for example model's forward function.
+        input_tensor:      Input at which gradients are evaluated, will be passed to forward_fn.
+        target_ind: Index of the target class for which gradients must be computed (classification only).
+        additional_forward_args: Additional input arguments that forward function requires.
+    """
+    with torch.autograd.set_grad_enabled(True):
+        outputs = _run_forward(
+            forward_fn, input_tensor, target_ind, additional_forward_args
+        )
+        assert outputs[0].numel() == 1, (
+            "Target not provided when necessary, cannot"
+            " take gradient with respect to multiple outputs."
+        )
+        grads = torch.autograd.grad(torch.unbind(outputs), input_tensor)
+    return grads[0]
+
+
+def _forward_layer_distributed_eval(
+    forward_fn: ModelType,
+    inputs: torch.Tensor,
+    layer: LayerType,
+    target_ind: Optional[TargetType] = None,
+    additional_forward_args: Any = None,
+    attribute_to_layer_input: bool = False,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    r"""
+    A helper function that allows to set a hook on model's `layer`, run the forward
+    pass and return layer result and optionally also the output of the forward function
+    depending on whether we set `attribute_to_layer_input` to True or False.
+    """
+    saved_layer: torch.Tensor
+
+    def pre_hook_wrapper(_):
+        def forward_pre_hook(_, input_tensor):
+            nonlocal saved_layer
+            saved_layer = input_tensor[0]
+            return saved_layer.clone()
+
+        return forward_pre_hook
+
+    def hook_wrapper(_):
+        def forward_hook(_, input_tensor, output):  # pylint: disable=unused-argument
+            nonlocal saved_layer
+            saved_layer = output
+            return saved_layer.clone()
+
+        return forward_hook
+
+    hook = None
```

Comment:
Missing type of `hook` variable.

Reply:
Hook is of type RemovableHandle which is not importable from pytroch. I can add Any, but does it make sense?

Reply:
I was able to import it:
```bash
Python 3.9.16 (main, Dec 21 2022, 11:19:13) 
[GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> from torch.utils.hooks import RemovableHandle
>>> type(RemovableHandle)
<class 'type'>
>>> RemovableHandle()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__() missing 1 required positional argument: 'hooks_dict'
>>> torch.__version__
'1.13.1+cu117'
>>>
```